server/fileupdater.mjs

```diff
@@ -7,56 +7,106 @@ export default function FileUpdater(state) {
   if(v > VERSION)
     throw Error(`File version ${v} is newer than the supported version ${VERSION}.`);
 
+  const globalProperties = computeGlobalProperties(state, v);
   for(const id in state)
-    updateProperties(state[id], v);
+    updateProperties(state[id], v, globalProperties);
 
   state._meta.version = VERSION;
   return state;
 }
 
-function updateProperties(properties, v) {
+function computeGlobalProperties(state, v) {
+  let globalProperties = {};
+  if (globalProperties.v12DropShadowAllowed = v < 10) {
+    for (const id in state) {
+      const properties = state[id];
+      if (properties.type == 'card' || properties.type == 'deck' || properties.type == 'pile') {
+        globalProperties.v12DropShadowAllowed = globalProperties.v12DropShadowAllowed &&
+            !hasPropertyCondition(properties, (properties) => {
+              return properties.parentChangeRoutine || properties.changeRoutine;
+            });
+      }
+      globalProperties.v12DropShadowAllowed = globalProperties.v12DropShadowAllowed &&
+          !hasPropertyCondition(properties, (properties) => {
+            return properties.parentGlobalUpdateRoutine || properties.globalUpdateRoutine;
+          });
+      if (!globalProperties.v12DropShadowAllowed)
+        break;
+    }
+  }
+  return globalProperties;
+}
+
+function hasPropertyCondition(properties, condition) {
+  if (typeof properties != 'object')
+    return false;
+  if (condition(properties))
+    return true;
+  if (properties.type) {
+    if (hasPropertyCondition(properties.onPileCreation, condition))
+      return true;
```

Comment:
Should this also check `onEnter` and `onLeave`?

Reply:
Done.